File: app/src/main/java/vn/loitp/app/activity/tutorial/rxjava2/AsyncTaskRxActivity.java
package vn.loitp.app.activity.tutorial.rxjava2;

import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.core.base.BaseFontActivity;
import com.core.utilities.LLog;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Single;
import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import kotlin.Unit;
import loitp.basemaster.R;
import vn.loitp.app.activity.tutorial.rxjava2.model.Bike;

public class AsyncTaskRxActivity extends BaseFontActivity implements View.OnClickListener {
    private TextView tv;
    private Test1 test1;
    @NonNull
    private CompositeDisposable compositeDisposable = new CompositeDisposable();
    private Disposable disposable;

    @Override

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        tv = (TextView) findViewById(R.id.tv);
        findViewById(R.id.bt_async_task).setOnClickListener(this);
        findViewById(R.id.bt_rx_1).setOnClickListener(this);
        findViewById(R.id.bt_rx_2).setOnClickListener(this);
        findViewById(R.id.bt_rx_3).setOnClickListener(this);
        findViewById(R.id.bt_rx_4).setOnClickListener(this);
    }

    @Override
    protected boolean setFullScreen() {
        return false;
    }

    @Override
    protected String setTag() {
        return getClass().getSimpleName();
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_asynctask_rx;
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.bt_async_task:
                if (test1 != null) {
                    test1.cancel(true);
                }
                test1 = new Test1();
                test1.execute();
                break;
            case R.id.bt_rx_1:
                if (disposable != null) {
                    disposable.dispose();
                }
                final MyRxTask1 myRxTask1 = new MyRxTask1(tv);
                disposable = myRxTask1.execute();
                compositeDisposable.add(disposable);
                break;
            case R.id.bt_rx_2:
                if (disposable != null) {
                    disposable.dispose();
                    return;
                }
                final MyRxTask2 myRxTask2 = new MyRxTask2(tv);
                disposable = myRxTask2.execute();
                compositeDisposable.add(disposable);
                break;
            case R.id.bt_rx_3:
                final Test2 test2 = new Test2(6);
                test2.apply();
                break;
            case R.id.bt_rx_4:
                testWithProgress();
                break;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (test1 != null) {
            test1.cancel(true);
        }
        compositeDisposable.clear();
    }

    private class Test1 extends AsyncTask<Void, Bike, List<Bike>> {

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            tv.setText("onPreExecute\n");
        }

        @Override
        protected List<Bike> doInBackground(Void... voids) {
            final List<Bike> bikeList = new ArrayList<>();
            for (int i = 0; i < 10; i++) {
                try {
                    Thread.sleep(1000);
                    final Bike bike = new Bike();
                    bike.setModel("Model " + i);
                    bike.setName("Name " + i);
                    bikeList.add(bike);
                    publishProgress(bike);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                    return null;
                }
            }
            return bikeList;
        }

        @Override
        protected void onProgressUpdate(Bike... values) {
            super.onProgressUpdate(values);
            final Bike bike = values[0];
            tv.append(bike.getName() + " - " + bike.getModel() + "\n");
        }

        @Override
        protected void onCancelled() {
            super.onCancelled();
            tv.append("onCancelled\n");
        }

        @Override
        protected void onPostExecute(List<Bike> bikeList) {
            super.onPostExecute(bikeList);
            tv.append("onPostExecute\n");
        }
    }

    private class Test2 {
        private int count;

        Test2(int count) {
            this.count = count;
        }

        public void apply() {
            tv.setText("Prev\n");

            /*PublishSubject publishSubject = new PublishSubject<>();
            publishSubject.observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.io())
                    .subscribe {

            }*/
            Single.fromCallable(() -> {
                List<Bike> bikeList = new ArrayList<>();
                for (int i = 0; i < count; i++) {
                    Thread.sleep(1000);
                    Bike bike = new Bike();
                    bike.setName("Name " + i);
                    bike.setModel("Model " + i);
                    bikeList.add(bike);
                }
                return bikeList;
            })
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new SingleObserver<List<Bike>>() {
                        @Override
                        public void onSubscribe(Disposable d) {
                            tv.append("onSubscribe\n");
                        }

                        @Override
                        public void onSuccess(List<Bike> bikes) {
                            tv.append("onSuccess\n");
                        }

                        @Override
                        public void onError(Throwable e) {
                            tv.append("onError\n");
                        }
                    });

        }
    }

    private void testWithProgress() {
        tv.setText("prev testWithProgress\n");
        compositeDisposable.clear();
        final int count = 10;
        final TestAsyncKotlin testAsyncKotlin = new TestAsyncKotlin(count);
        final Disposable disProgress = testAsyncKotlin.subscribeProgression(integer -> {
            LLog.d("loitptest", "Home -> subscribeProgression " + integer);
            tv.append("Home -> subscribeProgression " + integer + "\n");
            return Unit.INSTANCE;
        });
        final Disposable dis = testAsyncKotlin.apply(aBoolean -> {
                    LLog.d("loitptest", "Home -> 1 aBoolean: " + aBoolean);
                    tv.append("Home -> 1 aBoolean: " + aBoolean + "\n");
                    return Unit.INSTANCE;
                }, throwable -> {
                    LLog.d("loitptest", "Home -> 2 throwable: " + throwable.toString());
                    tv.append("Home -> 2 throwable: " + throwable.toString() + "\n");
                    return Unit.INSTANCE;
                }, () -> {
                    LLog.d("loitptest", "Home -> on disposed");
                    tv.append("Home -> on disposed\n");
                    return Unit.INSTANCE;
                },
                () -> {
                    LLog.d("loitptest", "Home -> 3 finished");
                    tv.append("Home -> 3 finished\n");
                    return Unit.INSTANCE;
                });
        compositeDisposable.add(disProgress);
        compositeDisposable.add(dis);
    }
}

File: app/src/main/java/vn/loitp/app/activity/tutorial/rxjava2/AsyncTaskRxActivity.kt
package vn.loitp.app.activity.tutorial.rxjava2

import android.annotation.SuppressLint
import android.os.AsyncTask
import android.os.Bundle
import android.view.View
import androidx.core.view.isVisible
import com.loitpcore.annotation.IsFullScreen
import com.loitpcore.annotation.LogTag
import com.loitpcore.core.base.BaseFontActivity
import com.loitpcore.core.utilities.LUIUtil
import io.reactivex.Single
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_asynctask_rx.*
import vn.loitp.app.R
import vn.loitp.app.activity.tutorial.rxjava2.model.Bike

@LogTag("AsyncTaskRxActivity")
@IsFullScreen(false)
class AsyncTaskRxActivity : BaseFontActivity(), View.OnClickListener {
    private var taskTest1: TaskTest1? = null
    private var disposable: Disposable? = null

    override fun setLayoutResourceId(): Int {
        return R.layout.activity_asynctask_rx
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setupViews()
    }

    private fun setupViews() {
        lActionBar.apply {
            LUIUtil.setSafeOnClickListenerElastic(
                view = this.ivIconLeft,
                runnable = {
                    onBaseBackPressed()
                }
            )
            this.ivIconRight?.setImageResource(R.color.transparent)
            this.viewShadow?.isVisible = true
            this.tvTitle?.text = AsyncTaskRxActivity::class.java.simpleName
        }
        btAsyncTask.setOnClickListener(this)
        btRx1.setOnClickListener(this)
        btRx2.setOnClickListener(this)
        btRx3.setOnClickListener(this)
        btRx4.setOnClickListener(this)
    }

    @Suppress("DEPRECATION")
    override fun onClick(view: View) {
        when (view) {
            btAsyncTask -> {
                taskTest1?.cancel(true)
                taskTest1 = TaskTest1()
                taskTest1?.execute()
            }
            btRx1 -> {
                disposable?.dispose()
                val myRxTask1 = MyRxTask1(textView)
                disposable = myRxTask1.execute()
                disposable?.let {
                    compositeDisposable.add(it)
                }
            }
            btRx2 -> {
                disposable?.dispose()
                val myRxTask2 = MyRxTask2(textView)
                disposable = myRxTask2.execute()
                disposable?.let {
                    compositeDisposable.add(it)
                }
            }
            btRx3 -> {
                val test2 = Test2(6)
                test2.apply()
            }
            btRx4 -> testWithProgress()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        taskTest1?.cancel(true)
        compositeDisposable.clear()
    }

    @Suppress("DEPRECATION")
    @SuppressLint("StaticFieldLeak")
    private inner class TaskTest1 : AsyncTask<Void, Bike, List<Bike>>() {
        @Deprecated("Deprecated in Java")
        @SuppressLint("SetTextI18n")
        override fun onPreExecute() {
            super.onPreExecute()
            textView.text = "onPreExecute"
        }

        @Deprecated("Deprecated in Java")
        override fun doInBackground(vararg voids: Void): List<Bike> {
            val bikeList: MutableList<Bike> = ArrayList()
            for (i in 0..9) {
                try {
                    Thread.sleep(1000)
                    val bike = Bike()
                    bike.model = "Model $i"
                    bike.name = "Name $i"
                    bikeList.add(bike)
                    publishProgress(bike)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                    return emptyList()
                }
            }
            return bikeList
        }

        @Deprecated("Deprecated in Java")
        override fun onProgressUpdate(vararg values: Bike) {
            super.onProgressUpdate(*values)
            val bike = values[0]
            textView.append("\n${bike.name} - ${bike.model}")
        }

        @Deprecated("Deprecated in Java")
        override fun onCancelled() {
            super.onCancelled()
            textView.append("\nonCancelled")
        }

        @Deprecated("Deprecated in Java")
        override fun onPostExecute(bikeList: List<Bike>) {
            super.onPostExecute(bikeList)
            textView.append("\nPostExecute")
        }
    }

    private inner class Test2(private val count: Int) {
        @SuppressLint("SetTextI18n")
        fun apply() {
            textView.text = "Prev"

            Single.fromCallable<List<Bike>> {
                val bikeList: MutableList<Bike> = ArrayList()
                for (i in 0 until count) {
                    Thread.sleep(1000)
                    val bike = Bike()
                    bike.name = "Name $i"
                    bike.model = "Model $i"
                    bikeList.add(bike)
                }
                bikeList
            }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SingleObserver<List<Bike?>?> {
                    override fun onSubscribe(d: Disposable) {
                        textView.append("\nonSubscribe")
                    }

                    override fun onSuccess(bikes: List<Bike?>) {
                        textView.append("\nonSuccess")
                    }

                    override fun onError(e: Throwable) {
                        textView.append("\nonError: $e")
                    }
                })
        }
    }

    @SuppressLint("SetTextI18n")
    private fun testWithProgress() {
        textView.text = "prev testWithProgress\n"
        compositeDisposable.clear()
        val count = 10
        val testAsyncKotlin = TestAsyncKotlin(count)
        val disProgress = testAsyncKotlin.subscribeProgression { integer: Int ->
            logD("Home -> subscribeProgression $integer")
            textView.append("Home -> subscribeProgression $integer\n")
        }
        val dis = testAsyncKotlin.apply(
            { aBoolean: Boolean ->
                logD("Home -> 1 aBoolean: $aBoolean")
                textView.append("Home -> 1 aBoolean: $aBoolean\n")
            }, { throwable: Throwable ->
                logD("Home -> 2 throwable: $throwable")
                textView.append("Home -> 2 throwable: $throwable\n")
            }, {
                logD("Home -> on disposed")
                textView.append("Home -> on disposed\n")
            }
        ) {
            logD("Home -> 3 finished")
            textView.append("Home -> 3 finished\n")
        }
        compositeDisposable.add(disProgress)
        compositeDisposable.add(dis)
    }
}
